### ISSUE TITLE
Avoid unnecessary Cargo rebuilds when developing

### ISSUE THREAD

**Issue by david-perez** [awslabs/smithy-rs#1412]
The goal of this task is to avoid unnecessary Cargo rebuilds of the generated crates to decrease development iteration cycles.

The command `./gradlew --info -P modules='simple' -P cargoCommands='test' codegen-server-test:build` code-generates the [`simple.smithy` model](https://github.com/awslabs/smithy-rs/blob/main/codegen-server-test/model/simple.smithy) using the `rust-server-codegen` plugin, places the generated crate under `codegen-server-test/build/smithyprojections/codegen-server-test/simple-rust-servercodegen`, and runs `cargo test` on it. This is a very common command that I run during development to ensure the changes I'm making result in a crate that compiles and does not break any protocol tests so far. On my `c5.4xlarge` machine, this command averages around 30 seconds. Around 12 seconds are spent in Cargo compiling the crate. This is expected, it has many dependencies.

However, if you run the command _twice_ (without making any changes to the codegen code), both invocations take the same time: Cargo recompiles the crate and its dependencies, even when the generated crate is identical. This is because Gradle deletes everything under `buildDir` when running the `generateSmithyBuild` task:

```
> Task :codegen-server-test:smithyBuildJar
Deleting stale output file: /local/home/davidpz/workplace/smithy-ws/src/SmithyRsSource/codegen-server-test/build/smithyprojections/codegen-server-test
```

So the files get recreated each time. [Cargo uses `mtime`](https://github.com/rust-lang/cargo/issues/6529) (among other factors) to determine whether it needs to recompile a unit. Indeed, running with `CARGO_LOG=cargo::core::compiler::fingerprint=trace` yields:

```
[2022-05-26T09:25:19Z DEBUG cargo::core::compiler::fingerprint] max dep mtime for "/local/home/davidpz/workplace/smithy-ws/src/SmithyRsSource/codegen-server-test/build/smithyprojections/codegen-server-test/simple/rust-server-codegen" is "/local/home/davidpz/workplace/smithy-ws/src/SmithyRsSource/target/debug/deps/libpretty_assertions-41528a536b021c4f.rmeta" 1653080703.105489628s
[2022-05-26T09:25:19Z INFO  cargo::core::compiler::fingerprint] fingerprint error for simple v0.0.1 (/local/home/davidpz/workplace/smithy-ws/src/SmithyRsSource/codegen-server-test/build/smithyprojections/codegen-server-test/simple/rust-server-codegen)/Test/TargetInner { ..: lib_target("simple", ["lib"], "/local/home/davidpz/workplace/smithy-ws/src/SmithyRsSource/codegen-server-test/build/smithyprojections/codegen-server-test/simple/rust-server-codegen/src/lib.rs", Edition2021) }
[2022-05-26T09:25:19Z INFO  cargo::core::compiler::fingerprint]     err: current filesystem status shows we're outdated
```

That is the _first_ "fingerprint error" it logs. I don't understand why it then logs other fingerprint errors for some of the dependencies, some because of `mtime`, others because `RUSTFLAGS` changed (none of which should happen, these artifacts [reside in the repo's `target` dir](https://github.com/awslabs/smithy-rs/blob/main/.cargo/config.toml) and do not change across invocations):

```
[2022-05-26T09:25:19Z DEBUG cargo::core::compiler::fingerprint] fingerprint at: /local/home/davidpz/workplace/smithy-ws/src/SmithyRsSource/target/debug/.fingerprint/proc-macro-error-attr-d37dbbcbe4fd4dbe/build-script-build-script-build
[2022-05-26T09:25:19Z INFO  cargo::core::compiler::fingerprint] fingerprint error for proc-macro-error-attr v1.0.4/Build/TargetInner { ..: custom_build_target("build-script-build", "/home/davidpz/.cargo/registry/src/github.com-1ecc6299db9ec823/proc-macro-error-attr-1.0.4/build.rs", Edition2018) }
[2022-05-26T09:25:19Z INFO  cargo::core::compiler::fingerprint]     err: RUSTFLAGS has changed: previously [], now ["-D", "warnings"]
[2022-05-26T09:25:19Z DEBUG cargo::core::compiler::fingerprint] fingerprint at: /local/home/davidpz/workplace/smithy-ws/src/SmithyRsSource/target/debug/.fingerprint/serde_json-2ba482ddd001d657/lib-serde_json
[2022-05-26T09:25:19Z INFO  cargo::core::compiler::fingerprint] fingerprint error for serde_json v1.0.81/Build/TargetInner { ..: lib_target("serde_json", ["lib"], "/home/davidpz/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_json-1.0.81/src/lib.rs", Edition2018) }
[2022-05-26T09:25:19Z INFO  cargo::core::compiler::fingerprint]     err: current filesystem status shows we're outdated
```

I wrote a simple patch to, before `generateSmithyBuild` runs, cache the hashes of all the crate's files. Before `cargo` is invoked, I then restore the `mtime`s to what they were if the hashes coincide:

```diff
diff --git a/codegen-server-test/build.gradle.kts b/codegen-server-test/build.gradle.kts
index d124d843..6d020eb4 100644
--- a/codegen-server-test/build.gradle.kts
+++ b/codegen-server-test/build.gradle.kts
@@ -57,6 +57,13 @@ tasks.register("generateSmithyBuild") {
                     codegenTests(properties, allCodegenTests)
                 )
             )
+
+        project.extra["previousBuildHashes"] = buildDir.walk()
+            .filter { it.name.endsWith(".rs") || it.name == "Cargo.toml" || it.name == "Cargo.lock" }
+            .map {
+                HashUtils.getCheckSumFromFile(it) to it.lastModified()
+            }
+            .toMap()
     }
 }

@@ -81,10 +88,39 @@ tasks.register<Exec>(Cargo.CHECK.toString) {
 tasks.register<Exec>(Cargo.TEST.toString) {
     workingDir("$buildDir/$workingDirUnderBuildDir")
     environment("RUSTFLAGS", defaultRustFlags)
+    environment("CARGO_LOG", "cargo::core::compiler::fingerprint=trace")
     commandLine("cargo", "test")
     dependsOn("assemble")
 }

+tasks.register("modifyMtime") {
+    description = "modify Rust files' `mtime` if the contents did not change"
+    val generateSmithyBuildTask = dependsOn("generateSmithyBuild")
+
+    doLast {
+        val previousBuildHashes: Map<String, Long> = project.extra["previousBuildHashes"] as Map<String, Long>
+        println("previousBuildHashes")
+        previousBuildHashes.forEach {
+            println(it)
+        }
+
+        println("currentBuildHashes")
+        buildDir.walk()
+            .filter { it.name.endsWith(".rs") || it.name == "Cargo.toml" }
+            .map {
+                HashUtils.getCheckSumFromFile(it) to it
+            }
+            .forEach { (currentHash, currentFile) ->
+                previousBuildHashes[currentHash]?.also { oldMtime ->
+                    println("Setting `mtime` of $currentFile to $oldMtime because its hash $currentHash remained unchanged after a rebuild.")
+                    currentFile.setLastModified(oldMtime)
+                }
+            }
+    }
+}
+
+tasks[Cargo.TEST.toString].dependsOn("modifyMtime")
+
 tasks.register<Exec>(Cargo.DOCS.toString) {
     workingDir("$buildDir/$workingDirUnderBuildDir")
     environment("RUSTDOCFLAGS", defaultRustDocFlags)
```

[HashUtils](https://gist.github.com/LongClipeus/84db46e7d9714f67c4cbc40a67c8be1e).

I verified that with this patch the `mtime`'s of the generated files do not change across command invocations using `stat -c %Y`. However, Cargo still claims "current filesystem status shows we're outdated". I looked into Cargo to learn what might be causing this but I didn't find anything shedding more light.

This is where the determination is made, it's the last cause for a rebuild it checks:

```rust
        if !self.fs_status.up_to_date() {
            bail!("current filesystem status shows we're outdated");
        }
```

[link](https://github.com/rust-lang/cargo/blob/39ad1039d9e3e1746177bf5d134af4c164f95528/src/cargo/core/compiler/fingerprint.rs#L972-L974)

```rust
   let result = new_fingerprint.compare(&old_fingerprint);
```

[link](https://github.com/rust-lang/cargo/blob/39ad1039d9e3e1746177bf5d134af4c164f95528/src/cargo/core/compiler/fingerprint.rs#L1654-L1654)

It's the _new_ fingerprint's `FsStatus` that is stale:

```rust
    /// Description of whether the filesystem status for this unit is up to date
    /// or should be considered stale.
    #[serde(skip)]
    fs_status: FsStatus,
```

[link](https://github.com/rust-lang/cargo/blob/39ad1039d9e3e1746177bf5d134af4c164f95528/src/cargo/core/compiler/fingerprint.rs#L549-L552)

A fingerprint's `FsStatus` begins being `Stale`:

```rust
    Ok(Fingerprint {
        rustc: util::hash_u64(&cx.bcx.rustc().verbose_version),
        target: util::hash_u64(&unit.target),
        profile: profile_hash,
        // Note that .0 is hashed here, not .1 which is the cwd. That doesn't
        // actually affect the output artifact so there's no need to hash it.
        path: util::hash_u64(path_args(cx.bcx.ws, unit).0),
        features: format!("{:?}", unit.features),
        deps,
        local: Mutex::new(local),
        memoized_hash: Mutex::new(None),
        metadata,
        config: config.finish(),
        compile_kind,
        rustflags: extra_flags,
        fs_status: FsStatus::Stale,
        outputs,
    })
```

[link](https://github.com/rust-lang/cargo/blob/39ad1039d9e3e1746177bf5d134af4c164f95528/src/cargo/core/compiler/fingerprint.rs#L1334-L1351)

And only after it's attempted to be updated:

```rust
        // Everything was up to date! Record such.
        self.fs_status = FsStatus::UpToDate { mtimes };
        debug!("filesystem up-to-date {:?}", pkg_root);
```

[link](https://github.com/rust-lang/cargo/blob/39ad1039d9e3e1746177bf5d134af4c164f95528/src/cargo/core/compiler/fingerprint.rs#L1096-L1098)

So we're reaching this line, but bailing out somewhere before we reach the above one (?):

```rust
        debug!(
            "max output mtime for {:?} is {:?} {}",
            pkg_root, max_path, max_mtime
        );
```

[link](https://github.com/rust-lang/cargo/blob/39ad1039d9e3e1746177bf5d134af4c164f95528/src/cargo/core/compiler/fingerprint.rs#L1032-L1035)

**Comment by david-perez** [comment]
Another issue is causing unnecessary crate rebuilds. We are sending `RUSTFLAGS=-D warnings` when invoking Cargo. However, a common thing to do after generating a crate is to open its contents in an editor. The editor's `rust-analyzer` compiles the crate and its dependencies without the `RUSTFLAGS` we had used earlier. The next time you rebuild the crate, Cargo claims the `err: RUSTFLAGS has changed: previously [], now ["-D", "warnings"]` shown above that I did not know how to explain, and recompiles everything again. If I instead generate a `.cargo/config.toml` containing these flags, and apply my `mtime` fixup patch above, the crate rebuilds go away and Cargo claims that everything is up to date!

Thanks @jonhoo.
